### DISCUSSION TITLE
Integer keys in my kedro config can't be logged using kedro-mflow

### DISCUSSION THREAD

**Discussion by Jarathael** [Galileo-Galilei/kedro-mlflow#224] [Q&A] [answered]
Hi !
I'm using this plugin for a machine learning project using kedro at my job.
I ended up with a weird situation: basically in `base/parameters/train.yml` I defined a bunch of parameters among of which I have the type of model I want to use and the parameters I want to be defined. In my code, the model is instantiated based on the model type and by passing directly the dictionnary of parameters to the init. However, the model I tried to instantiate was `sklearn.ensemble.RandomForestClassifier`. One of its parameters his `class_weight` which can be a dictionnary. The problem is that scikit-learn need a dictionnary with integer key. And the `flatten_dict` function defined in `kedro-mlflow` plugin is not taking this into account when flattening the parameters dictionnary.

Ideally I would want the whole dictionnary to be recorded since it is ONE parameter. But I don't think it would be easy to detect such a case anyway. However what I would like, at least, is for the `flatten_dict` to take into account that keys can be something else that strings and then cast them as strings in:
https://github.com/Galileo-Galilei/kedro-mlflow/blob/cb3ef40640e09ee018a4226935868181354abde8/kedro_mlflow/framework/hooks/node_hook.py#L137

How can I possibly do that directly into my project ? Is it possible to redefine this function in my project so the modified version is used instead ?

Thank you in advance !

**Comment by Galileo-Galilei** [comment]
Hi, glad to see that you are using the plugin!

I have a few questions regarding your issue:
- what happens if you have the ``flatten_dict_params`` set to ``False`` in your ``mlflow.yml``? I guess that should not have any error and the the whole dictionnary should be recorded as one single parameters as you require.
- Can you share a minimal reproducible example as code (say the ``parameters.yml`` with the faulty parameters, and a very small function that reproduces the problem you describe?), as well as your ``python`` / ``kedro`` / ``mlflow`` / ``kedro-mlflow`` versions, and the error message (something like ``TypeError: unsupported operand type(s) for +: 'int' and 'str'`` I guess)?
- Can you reinstall from this branch and tell me if it fixes the bug?

```bash
pip uninstall kedro-mlflow
pip install git+https://github.com/Galileo-Galilei/kedro-mlflow.git@dict-params-int-keys
``` 

Do not forget to uninstall kedro-mlflow first, else the fix will not be reinstalled because they have the same version number.

**Reply by Jarathael** [reply]
If i set `flatten_dict_params` to `False` in `mlflow.yml` it does not throw any error. But I don't want that.  As to give you an idea here an extract of the config file I was talking about:
```
train:
  ...
  model:
    type: sklearn.ensemble.RandomForestClassifier
    params:
      random_state: 9834
      n_jobs: -1
      class_weight: {0: 1, 1: 4}
    cv_params:
      ...
  ...
```

I have several nodes using this all param dictionary. Here is an example:
```
node(
    func=node_train,
    inputs=["features_train", "target_train", "params:train"],
    outputs=["sklearn_model", "cv_results"],
    name="train_node",
)
``` 

As I said, in this node I can instantiate a model based on its `type` (I load the corresponding sklearn class) and use a starred expression on `train.model.params` to instantiate my model:
```
def load_sklearn_class(c_type):
    if not c_type.startswith("sklearn"):
            c_type = ".".join(["sklearn", c_type])
            self.logger.warning(f"Trying to load {c_type} from 'sklearn' implicitly.")
    module_name, class_name = c_type.rsplit(".", 1)
    try:
        o = getattr(importlib.import_module(module_name), class_name)
        return o
    except AttributeError as e:
        raise e

model = load_sklearn_class(model_config["type"])(model_config["params"])
```

The message error I get is: `TypeError: must be str, not int`

My versions:
```
python: 3.12.6
kedro: 0.17.3
mlflow: 1.19.0
kedro-mlflow: 0.7.2
```

This branch fixes the bug :)


Also in an ideal world I would like a record like that `train.model.params.class_weight: {0: 1, 1: 4}` but regarding my config it seems to be quite difficult to do it automatically I guess.

**Comment by Galileo-Galilei** [comment, marked as answer]
Thank you for the details. I had understood the general idea but reading the code is always more precise. 

I'll merge the branch on master within 2 weeks maximum (once I have added a test for this use case) and release it on Pypi soon after that. 

Regarding your ideal logging format, I understand why this would be more readable for you, but it is very hard to find how to do it in a general way, since in most situations you want to flatten the entire dict. I could eventually modify the ``flatten_dict`` function to either: 
- add the possibility to flatten only until a given depth
- or avoid flattening the last "dict" in a nested structure (your use case)
but it does not bring much value and I have other priorities and more important features on top of my list right now. I'd be glad to accept a PR for this though, so feel free to take it ;)

**Reply by Jarathael** [reply]
You're welcome !

Thank you for this hotfix :)

Yes, I know it's hard to find a way to do it in a general way. But I guess having the dict completely flatten is not that bad eventually if I want to compare different runs. So it's fine like this !

Thank you again.
